src/Utils.ts

```diff
@@ -986,12 +986,129 @@ function setDummyInputDebugValue(
     );
 }
 
+export class DummyInputObserver implements Types.DummyInputObserver {
+    private _win?: GetWindow;
+    private _offsetsQueue: (() => void)[] = [];
+    private _offsetsTimer?: number;
+    private _domChangedMap: WeakMap<HTMLElement, true> = new WeakMap();
+    private _domChangedTimer?: number;
+    private _dummyMap: WeakMap<HTMLElement, () => void> = new WeakMap();
+    private _dummies: WeakRef<HTMLElement>[] = [];
+    domChanged?(parent: HTMLElement | null): void;
+
+    constructor(win: GetWindow) {
+        this._win = win;
+    }
+
+    add(dummy: HTMLElement | undefined, callback: () => void): void {
+        if (dummy) {
+            const dummyMap = this._dummyMap;
+            const cb = dummyMap.get(dummy);
+
+            if (!cb) {
+                this._dummies.push(new WeakRef(dummy));
+                this.domChanged = this._domChanged;
+            }
+
+            dummyMap.set(dummy, callback);
+        }
+    }
+
+    remove(dummy: HTMLElement | undefined): void {
+        if (dummy) {
+            const dummyMap = this._dummyMap;
+            const cb = dummyMap.get(dummy);
+
+            if (cb) {
+                dummyMap.delete(dummy);
+
+                const dummies = this._dummies;
+
+                for (let i = dummies.length; i--; ) {
+                    const el = dummies[i].deref();
+
+                    if (el) {
+                        if (el === dummy) {
+                            dummies.splice(i, 1);
+                            break;
+                        }
+                    } else {
+                        dummies.splice(i, 1);
+                    }
+                }
+
+                if (dummies.length === 0) {
+                    delete this.domChanged;
+                }
+            }
+        }
```

Comment:
```suggestion
        if (!dummy) {
          return
        }
        const dummyMap = this._dummyMap;
        const cb = dummyMap.get(dummy);

        if (!cb) {
          return
        }

        dummyMap.delete(dummy);

        const dummies = this._dummies;

        for (let i = dummies.length; i--; ) {
            const el = dummies[i].deref();

            if (el === dummy) {
              dummies.splice(i, 1);
              break;
            }
            if (el === undefined) {
              dummies.splice(i, 1);
            }
        }

        if (dummies.length === 0) {
            delete this.domChanged;
        }
```

avoid adding another level of nesting by short returning (the less nested the easier to read)